pkg/controller/deployment/util/deployment_util.go

```diff
@@ -714,6 +714,17 @@ func GetAvailableReplicaCountForReplicaSets(replicaSets []*apps.ReplicaSet) int3
 	return totalAvailableReplicas
 }
 
+// GetTerminatingReplicaCountForReplicaSets returns the number of terminating pods for all replica sets.
+func GetTerminatingReplicaCountForReplicaSets(replicaSets []*apps.ReplicaSet) int32 {
+	terminatingReplicas := int32(0)
+	for _, rs := range replicaSets {
+		if rs != nil {
+			terminatingReplicas += rs.Status.TerminatingReplicas
```

Comment:
if some replicas do not have the field set in status, I think we have to return nil here (we can't total them)

Reply:
I think we should not let unsynced RSs let the deployment status sink to 0. And show terminating replicas at least from the last known good state. We will update it to the latest once the replica set controller syncs the state to the latest.

We also have a delay until the replica set controller picks up a new terminating pod.

Reply:
> I think we should not let unsynced RSs let the deployment status sink to 0

If there are *any* RSs that don't have the status field set, we cannot put anything other than nil in the deployment status

Reply:
I think that would be confusing for the users and the expectation is that the status counts all the terminating pods that are known at a specific time.

Example:

2RS running with 2+1 terminating pods -> `.status.terminatingReplicas == 3`

3RS running with 2+1 terminating pods and one new unsynced RS -> `.status.terminatingReplicas == nil`


3RS running with 2+1 terminating pods fully synced -> `.status.terminatingReplicas == 3`


Anyone inferring information from the deployment this could get a buggy/racy behavior.

Reply:
> Anyone inferring information from the deployment status could get a buggy/racy behavior.

We're expecting the *deployment controller* to infer and act on that information, right?

Reply:
We do count the `.status.replicas`, `.status.readyReplicas`, etc. from the same set of ReplicaSets when calculating the status for `.status.terminatingReplicas`.

Reply:
> We're expecting the deployment controller to infer and act on that information, right?

yes we do, and we should start new replicas in the meantime, before the new RS get synced

Reply:
I mean, it is acknowledging the terminating pods it knows about and not spinning up more pods. E.g. during scaling.

Realistically to achieve the RS status to be empty:
- either there is a new unsynced RS, which did not have a chance to start any pods yet and thus it also does not have any terminating pods
- somebody could also creates an extra pod manually when RS is down, but this can happen also for already running RS and we would not be calculating that correctly either
- somebody deletes the status of RS or/and stops the RS controller - this is already a hacky behavior